Re-write LocalStateQuery client logic to eliminate acquire failures

Anviking opened the pull request:
# Issue Number

ADP-647, #2320

<!-- Put here a reference to the issue that this PR relates to and which requirements it tackles. Jira issues of the form ADP- will be auto-linked. -->


# Overview

- [x] Allow `send`-ing a composition of queries against a single acquired point, not just one.
- [x] Makes acquire failures practically impossible
- [x] Re-add multi-era support with reduced boilerplate
- [x] Some polish still needed
- [x] Re-add tracing of query times (less granular than before, but done)

# Comments

Pretty sure this _is_
- eliminating acquire failures
- the right direction

But also
- Might introduce a new set of problems

<!-- Additional comments or screenshots to attach if any -->

## Failures

I have run tests locally a lot on this branch. One new failure I _occasionally_ see is
```
  src/Test/Integration/Framework/DSL.hs:1797:7:
  1) API Specifications, SHELLEY_STAKE_POOLS, STAKE_POOLS_JOIN_01rewards - Can join a pool, earn rewards and collect them
       expected a successful response but got an error: DecodeFailure "{\"code\":\"created_invalid_transaction\",\"message\":\"That's embarrassing. It looks like I've created an invalid transaction that could not be parsed by the node. Here's an error message that may help with debugging: HardForkApplyTxErrFromEra S (S (S (Z (WrapApplyTxErr {unwrapApplyTxErr = ApplyTxError [LedgerFailure (DelegsFailure (WithdrawalsNotInRewardsDELEGS (fromList [(RewardAcnt {getRwdNetwork = Mainnet, getRwdCred = KeyHashObj (KeyHash \\\"9c0ff007dd21bbf24960bd12ae4009efb8cad076228ef1a54c7b5dbe\\\")},Coin 7010064794)])))]}))))\"}"
       While verifying (Status {statusCode = 500, statusMessage = "Internal Server Error"},Left (DecodeFailure "{\"code\":\"created_invalid_transaction\",\"message\":\"That's embarrassing. It looks like I've created an invalid transaction that could not be parsed by the node. Here's an error message that may help with debugging: HardForkApplyTxErrFromEra S (S (S (Z (WrapApplyTxErr {unwrapApplyTxErr = ApplyTxError [LedgerFailure (DelegsFailure (WithdrawalsNotInRewardsDELEGS (fromList [(RewardAcnt {getRwdNetwork = Mainnet, getRwdCred = KeyHashObj (KeyHash \\\"9c0ff007dd21bbf24960bd12ae4009efb8cad076228ef1a54c7b5dbe\\\")},Coin 7010064794)])))]}))))\"}"))

  To rerun use: --match "/API Specifications/SHELLEY_STAKE_POOLS/STAKE_POOLS_JOIN_01rewards - Can join a pool, earn rewards and collect them/"
```
Important part is `WithdrawalsNotInRewardsDELEGS`. So seems we are not aware of the rewards having already been spent. Maybe this PR makes the rewards slower to update, somehow, not sure.

Edit: digging through my notes, I _have_ seen this failure on another branch — once. I think this PR makes it more likely to occur, but not be inherently related, then.

Anviking committed:
Re-write local state query client code with LSQ monad

Goals
- Make AcquireFailures impossible (for wallet API users)
- Eliminate integration test flakiness
- Reduce boilerplate

by adding a monad for composing queries to be run against any, but the
same, node tip.

* wip: leverage STM retrying

This adopts the state observation pattern Duncan described in a recent
STM seminar.

We now use pure TVar's. Not TMVar or Chan. We can still efficiently
observe for changes by using `guard`.

This seems to make the tests much greener. Hoping this fixes several
flaky tests on master.

* Reduce boilerplate, and re-add support for past eras

Anviking commented on lib/core/src/Ouroboros/Network/Client/Wallet.hs:
```diff
@@ -475,62 +479,99 @@ type LocalStateQueryResult state = Either AcquireFailure state
+data LSQ block (m :: * -> *) a where
+    LSQPure :: a -> LSQ block m a
+    LSQBind :: LSQ block m a -> (a -> LSQ block m b) -> LSQ block m b
+    LSQry :: (Query block res) -> LSQ block m res
```
👆 The LSQ monad that makes this all possible

Anviking commented on lib/core/src/Ouroboros/Network/Client/Wallet.hs:
```diff
@@ -475,62 +479,99 @@ type LocalStateQueryResult state = Either AcquireFailure state
+                go
+            else do
+                --respond (Left failure)
+                -- TODO: Fail
```
I don't think clients should have to care about acquire failures. If we can't get a tip acquired with infinite retries, then something is truly wrong.

(Should in that case have infinite retries though)

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -538,82 +546,17 @@ withNetworkLayerBase tr np addrInfo (versionData, _) action = do
-        queryStakeDistribution pt = \case
-            AnyCardanoEra ShelleyEra -> do
-                let cmd = CmdQueryLocalState pt (QueryIfCurrentShelley Shelley.GetStakeDistribution)
-                result <- handleQueryResult "GetStakeDistribution"
```
The tracing would need to be re-added when _interpreting_ the LSQ monad.

Anviking replied:
Tricky, cf https://github.com/input-output-hk/cardano-wallet/pull/2449#discussion_r559741621

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -820,98 +766,25 @@ mkTipSyncClient tr np localTxSubmissionQ onTipUpdate onPParamsUpdate onInterpret
+                    mb <- LSQry $ QueryAnytimeMary GetEraStart
+                    sp <- fmap (slottingParametersFromGenesis . getCompactGenesis)
+                        <$> LSQry (QueryIfCurrentMary Shelley.GetGenesisConfig)
+                    pp' <- LSQry $ QueryIfCurrentMary Shelley.GetCurrentPParams
```
This is actually readable!

(but, ok some complexity needs to be re-added)

Anviking replied:
added `shelleyBased` and `byronOrShelleyBased` helpers

Anviking commented on lib/core/src/Ouroboros/Network/Client/Wallet.hs:
```diff
@@ -475,62 +479,99 @@ type LocalStateQueryResult state = Either AcquireFailure state
+                        newPt <- getTip
+                        if oldPt == newPt
+                        -- FIXME: Not the most elegant approach of waiting for a
+                        -- new tip.
```
Should probably pass in a `TChan m ((Point block, AnyCardanoEra))`

Edit: absolutely not! Then we might get old values of the tip.

Anviking replied:
Duncan's STM seminar would suggest that we should use a plan `TVar` and retry inside STM!

KtorZ commented on lib/core/src/Ouroboros/Network/Client/Wallet.hs:
```diff
@@ -131,8 +151,10 @@ import qualified Ouroboros.Network.Protocol.LocalStateQuery.Client as LSQ
     :: forall m block era. (Monad m)
     => (block -> era)
-    -> (Maybe era -> Tip block -> m ())
+    -> (era -> Tip block -> m ())
```
I am afraid this may lead to nasty errors. The reason why I initially used a `Maybe` here was to stil allow `Tip` updates. This is perhaps confusing and we may consider having 2 distinct callbacks then: one for the `Tip` and one for the `era`.

Anviking replied:
What kind of nasty errors? It simplifies things, and I didn't see how it could be bad.

That said, now with `STM` `guard` it might be pretty ok to keep the `Maybe`… if we guard for `Just` values on the era.

KtorZ replied:
If we rollback, it means that for a while, we'll be using a tip that is in the future.

Anviking replied:
Should be done -> 30c569d

KtorZ commented on lib/core/src/Ouroboros/Network/Client/Wallet.hs:
```diff
@@ -475,62 +494,97 @@ type LocalStateQueryResult state = Either AcquireFailure state
+
+instance Applicative (LSQ block m) where
+    pure  = LSQPure
+    (<*>) = ap
```
The fact that `ap` is defined in terms of `Monad` is utterly confusing to me :smile: ... I guess this is because I started with PureScript and not Haskell, where everything comes from `Applicative` whereas Haskell usually defines things from `Monad` and downwards.

KtorZ commented on lib/core/src/Ouroboros/Network/Client/Wallet.hs:
```diff
@@ -475,62 +494,97 @@ type LocalStateQueryResult state = Either AcquireFailure state
-            LSQ.SendMsgAcquire pt (clientStAcquiring query respond)
+    clientStIdle = do
+        cmd <- awaitNextCmd
+        (era, pt) <- second getTipPoint <$> atomically (readTVar tipVar)
```
Seems like this is always using `Point block` instead of `Tip block`, why not store it as such then and avoid all the `getTipPoint` calls.

Anviking replied:
Can't `fmap` over `TVar`s, but should switch to taking `STM (AnyCardanoEra, Point block)` anyway, such that the read-onlyness is enforced.

Anviking replied:
Changed to `STM (AnyCardanoEra, Point block)` where possible.

KtorZ commented on lib/core/src/Ouroboros/Network/Client/Wallet.hs:
```diff
@@ -475,62 +494,97 @@ type LocalStateQueryResult state = Either AcquireFailure state
+        , recvMsgFailure = \_failure -> do
+            (newEra, newPt) <- second getTipPoint <$> (atomically $ do
+                t <- readTVar tipVar
+                guard (getTipPoint (snd t) /= oldPt)
```
So, in case the points previously known is different, the transaction is retried until it gets updated?

Anviking replied:
Yes, that's how I understood it from Duncan's talk, and it seems to work.

KtorZ replied:
I see that the `Alternative` instance for `STM` is defined as `empty = retry` so ... yes, I guess it works :sweat_smile:

Anviking replied:
And the retrying is efficient in some sense, because it only retries when a previously read value has changed

KtorZ commented on lib/core/src/Ouroboros/Network/Client/Wallet.hs:
```diff
@@ -475,62 +494,97 @@ type LocalStateQueryResult state = Either AcquireFailure state
+                t <- readTVar tipVar
+                guard (getTipPoint (snd t) /= oldPt)
+                return t)
+            pure $ LSQ.SendMsgAcquire newPt (clientStAcquiring qry newPt newEra (retries + 1))
```
This is counting retries, what for? Is the goal to cap that at some point?

Anviking replied:
Initially I capped it, but now it is not capped and this should be removed. (unless we'd want to trace the retries, but 🤷‍♂️ )

KtorZ commented on lib/core/src/Ouroboros/Network/Client/Wallet.hs:
```diff
@@ -475,62 +494,97 @@ type LocalStateQueryResult state = Either AcquireFailure state
+              pure $ cont res
+          go (LSQBind ma f) cont = go ma $ \a -> do
+              go (f a) $ \b -> cont b
+          go (LSQEra) cont = cont era
```
Hmmm. The continuation above will `respond`, that is, push a message into the response queue. This means that it is possible for callers to create "ill-programs" that will generate many responses in the queue. For instance `LsqPure a >> LsqPure a` will yield two messages. That's pretty bad as we may both create a memory leak but also fill the queue with responses which may no longer correspond to queries. 

Unless I am missing something?

Anviking replied:
Ah, now I get what you mean. No — I don't think so.

The final `respond` will only be called by the top-level `go`.

```
LSQPure a >> LSQPure a
== LSQBind (LSQPure a) (\_ -> LSQPure a)
```

which should evaluate something like this:
```
go ma $ \a -> go (f a) $ \b -> cont b
go (LSQPure a) $ \a -> go (f a) $ \b -> cont b
go (f a) $ \b -> cont b
go (LSQPure a) $ \b -> cont b
cont a
respond a
```

Anviking replied:
Could perhaps abstract the run-function to the top level and test it — but not sure how, on the top off my head.

Anviking replied:
Also: The danger I see is if the queries contain a partial function. If it throws, I think the entire state machine will come to a halt.

Anviking replied:
Luckily the entire wallet crashes in that case:
```
[cardano-wallet.network:Notice:282] [2021-01-18 19:34:23.67 UTC] Query stakePoolsSummary took 19.62426s (too slow)
cardano-wallet: ExceptionInLinkedThread ThreadId 12 todo
CallStack (from HasCallStack):
  error, called at src/Cardano/Wallet/Shelley/Network.hs:502:18 in cardano-wallet-2021.1.12-3Kq63VUOZn92WgCVghwpjO:Cardano.Wallet.Shelley.Network
  withNetworkLayerBase, called at src/Cardano/Wallet/Shelley/Network.hs:294:5 in cardano-wallet-2021.1.12-3Kq63VUOZn92WgCVghwpjO:Cardano.Wallet.Shelley.Network
  withNetworkLayer, called at src/Cardano/Wallet/Shelley.hs:274:9 in cardano-wallet-2021.1.12-3Kq63VUOZn92WgCVghwpjO:Cardano.Wallet.Shelley
```
So should be fine.

Anviking commented on lib/core/src/Ouroboros/Network/Client/Wallet.hs:
```diff
@@ -475,62 +479,103 @@ type LocalStateQueryResult state = Either AcquireFailure state
+          go (LSQPure a) cont = cont a
+          go (LSQry qry) cont = LSQ.SendMsgQuery qry $ LSQ.ClientStQuerying $ \res -> do
+              pure $ cont res
+              -- TODO: It would be nice to trace the time it takes to run the
```
This does not appear to be straight forward to do. There is no opportunity to run `IO` _before_ the query, but _after_ the point is acquired.

Anviking replied:
I re-added top-level `bracketQuery` in the network layer that time the full `send` action (includes acquiring, and all the included queries).

Anviking committed:
Revert to allowing the era to be Nothing

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -985,75 +841,43 @@ newRewardBalanceFetcher
+                        $ Set.map toStakeCredential accounts))
+
+        -- FIXME: Re-add trace
+        --liftIO $ traceWith tr $ MsgAccountDelegationAndRewards deleg rewardAccounts
```
☝️

Anviking replied:
Fixed in 8f00b12bef91b08cf007ab33d0deb06917c1dd19

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -317,43 +309,39 @@ withNetworkLayerBase tr np conn (versionData, _) action = do
 
+    -- Alternative read-action which waits for the era to be Just and uses Point
+    -- instead of Tip.
+    let readNodeEraAndTip' = readNodeEraAndTip >>= \case
```
Could possibly converge on some neat api of `STM` actions for node-tip state, defined in the tip sync client, but leaving it like this for now.

KtorZ replied:
On a small stylistic note, this logic is quite specific to the `currentNodeEra` handler so I'd simply leave in there rather than defining a top-level let binding.

Anviking committed:
Re-add MsgAccountDelegationAndRewards trace

Anviking committed:
Tweak comments

KtorZ commented on lib/core/src/Ouroboros/Network/Client/Wallet.hs:
```diff
@@ -475,62 +475,109 @@ type LocalStateQueryResult state = Either AcquireFailure state
 --
+-- NOTE: Using AnyCardanoEra arguably goes against the grain of the abstract
+-- block type. We might be able to use the (Header block) as replacemet for
+-- (AnyCardanoEra, Tip block).
```
I think this is a `FIXME`, but the one above is `NOTE` :)

rvl replied:
I fixed the state diagram.

KtorZ commented on lib/core/src/Ouroboros/Network/Client/Wallet.hs:
```diff
@@ -475,62 +475,109 @@ type LocalStateQueryResult state = Either AcquireFailure state
+                t <- getTip
+                guard (((snd t) /= oldPt))
+                return t
+            pure $ LSQ.SendMsgAcquire newPt (clientStAcquiring qry newPt newEra)
```
Did you check with the network team whether it is better to `ReAcquire` or `Acquire` in this situation? I am still not 100% sure about the semantic we should expect from the state query protocol in this regard.

rvl replied:
I was just looking at the LocalStateQuery types in ouroboros-network, and there is only one option after MsgFailure, and that's MsgReAcquire. In the Acquired state, the shortest way to move the state to another block is with MsgReAcquire.

Anviking commented on lib/core/src/Ouroboros/Network/Client/Wallet.hs:
```diff
@@ -475,62 +475,109 @@ type LocalStateQueryResult state = Either AcquireFailure state
+        LSQ.SendMsgRelease (respond res >> clientStIdle)
+            -- We /could/ read all LocalStateQueryCmds from the TQueue, and run
+            -- them against the same tip, if re-acquiring takes a long time. As
+            -- of Jan 2020, it seems like queries themselves take significantly
```
```suggestion
            -- of Jan 2021, it seems like queries themselves take significantly
```
Oops, wrong year!

KtorZ replied:
Living in the past :p

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -632,13 +547,16 @@ withNetworkLayerBase tr np conn (versionData, _) action = do
+                    return tip
+                cb (toBlockHeader . snd $ tip)
+                go tip
+        go (Just $ AnyCardanoEra ByronEra, TipGenesis )
```
we are loosing the `MsgWatcherUpdate` trace from this new approach.

Anviking replied:
👍 thanks

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -804,165 +723,102 @@ mkTipSyncClient
+                AnyCardanoEra ShelleyEra ->
+                    LSQry $ QueryAnytimeShelley GetEraStart
+                _ ->
+                    LSQry $ QueryAnytimeMary GetEraStart
```
Hmm. This seems wrong (not that this PR is changing anything but we should look into this).

Anviking replied:
Yes. I know. I'm not sure what it's supposed to do. When we change `hardfork_at` to return all eras should make it right though.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -804,165 +723,102 @@ mkTipSyncClient
+        -- NOTE 2: interpeter is updated every block. This is far more often than
+        -- necessary.
+        sendAsync localStateQueryQ $ SomeLSQ queryParams onPParamsUpdate'
+        sendAsync localStateQueryQ $ SomeLSQ queryInterpreter onInterpreterUpdate
```
Arguably, we only need those once per epoch.

Anviking replied:
It would be cool to add some logic for that. But while it would probably be easy, it would probably not  be completely trivial either.

Anviking replied:
For instance, we can't observe the node tip `STM` and filter out the epoch-changes, since we're not guaranteed to see all intermediate states with `STM`.

In that case we'd probably need the tip-sync client to have specific logic for running actions on epoch-boundaries.

KtorZ reviewed (approved):
Looks okay, although I admit being mildly scared by the number of changes in the network layer. Yet, I think it's going in the right direction and it makes the code simpler to reason about anyway; so new problems should also be easier to catch.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -985,75 +841,45 @@ newRewardBalanceFetcher
   where
     fetch
-        :: (AnyCardanoEra, Tip (CardanoBlock StandardCrypto))
+        :: (Maybe AnyCardanoEra, Tip (CardanoBlock StandardCrypto))
```
Hm. This `Tip` is only used for logging. And the log message might lie, since it is not necessarily the tip we're going to use.

Anviking replied:
I removed it from the trace.

Anviking committed:
Refactor newRewardBalanceFetcher

Anviking committed:
Remove tip from MsgGetRewardAccountBalance

The logged tip would not necessarily be the tip used to make the query,
which could be misleading.

Anviking committed:
Re-add MsgWatcherUpdate trace

Anviking committed:
Fixup year in comment

rvl committed:
Correct LocalStateQuery state diagram

Anviking committed:
Simplify LSQ by removing getTip

Simplify the local state query client by having it acquire the /latest/
tip (`Nothing`), instead of asking for a specific tip from the tip sync
client.

To know the era, we now need to send a query. But this arguably just
makes things neater. It also resolves a problem from the previous
commits, where LSQ would block on wallet startup until the node changed
tip (could be on the order of 10s), since the tip sync client only had
received a MsgRollback without information of the current era.

Anviking committed:
Add back-pressure to on-tip queries

If the tip changes alot and fast, we don't want to fill up the local
state query queue with multiple identical requests.

With this commit, the PParams and Interpreter queries have to succeed
before we potentially re-run them on a new tip.

Anviking committed:
Remove out-of-date comment
